tests/performance/resourcecollector/resource_collector.go

```diff
@@ -0,0 +1,282 @@
+		var err error
+		ep, err = rc.VirtClient.CoreV1().Endpoints(namespace).Get(context.Background(), prometheusService, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred(), "failed to retrieve Prometheus endpoint")
+
+		if len(ep.Subsets) == 0 || len(ep.Subsets[0].Addresses) == 0 {
+			return false
+		}
+		return true
+	}, 10*time.Second, time.Second).Should(BeTrue())
+
+	for _, port := range ep.Subsets[0].Ports {
+		if port.Name == "web" {
+			rc.ServicePrometheusPort = port.Port
+		}
+	}
+	Expect(rc.ServicePrometheusPort).ToNot(Equal(0), "could not get Prometheus port from endpoint")
+}
+
+// SetToken returns true if the token was found, otherwise, it means that Prometheus could not be found
+func (rc *ResourceCollector) SetToken() bool {
+	secrets, err := rc.VirtClient.CoreV1().Secrets(namespace).List(context.Background(), metav1.ListOptions{})
+	Expect(err).ToNot(HaveOccurred())
+	for _, secret := range secrets.Items {
+		if strings.Contains(secret.Name, prometheusSecret) {
+			token := string(secret.Data["token"])
+			rc.Token = "Bearer " + token
+			return true
+		}
+	}
+	return false
+}
+
+// CreatePromPortForward set up port forwarding for Prometheus service")
+func (rc *ResourceCollector) CreatePromPortForward() {
+	portMapping := fmt.Sprintf("%d:%d", localPrometheusPort, rc.ServicePrometheusPort)
+	_, kubectlCmd, err := tests.CreateCommandWithNS(namespace, "kubectl", "port-forward", "service/"+prometheusService, portMapping)
+	Expect(err).ToNot(HaveOccurred())
+
+	err = kubectlCmd.Start()
+	Expect(err).ToNot(HaveOccurred())
+}
+
+// Query queries Prometheus API endpoint for a given metric
+func (rc *ResourceCollector) Query(query string) (prom_model.Value, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	res, warnings, err := rc.Client.Query(ctx, query, time.Now())
+	if err != nil {
+		fmt.Printf("FATAL: Error querying Prometheus: %v\n", err)
+	}
+	if len(warnings) > 0 {
+		fmt.Printf("WARN: %v\n", warnings)
+	}
+	return res, err
+}
+
+// Query queries Prometheus API endpoint for a given metric
+func (rc *ResourceCollector) QueryToValues(sample prom_model.Value) []float64 {
+	var values []float64
+	for _, s := range sample.(model.Vector) {
+		values = append(values, float64(s.Value))
+	}
+	return values
+}
+
+// Query queries Prometheus API endpoint for a given metric
+func (rc *ResourceCollector) QueryToValuesAgg(sample prom_model.Value, agg Agg) float64 {
+	var values []float64
+	for _, s := range sample.(model.Vector) {
+		values = append(values, float64(s.Value))
+	}
+
+	total := 0.0
+	for _, v := range values {
+		total += v
+	}
+	if agg == AVG {
+		return total / float64(len(values))
+	}
+	return total
+}
+
+// GetApiRequestDuration queries Prometheus API endpoint for the kubevirt API REST request duration, it returns the 99-percentile of the latencies of APIS
+func (rc *ResourceCollector) GetApiRequestDuration(url string) (prom_model.Value, error) {
```

Comment:
We definitely want to have this information exported, but I don't see you wiring up many of these functions to anything.  Do these helper functions need to be a part of this PR?

Reply:
Humm, it is a good point! I forgot to include an API request latency verification...

The plan is to actually verify the API latency and fail the test if the latency is much higher than expected.
Also, note that, I am verifying only the calls related to VMI creation, there are other calls that I am ignoring for now...

So, I could remove this, or include the latency verification that is missing....

Reply:
I think there are a few ideas encapsulated in this PR and each of them are important.  I'd like to break down what I think you are trying to accomplish into individual designs so each idea gets the attention it deserves.  Here's my understanding:
1) Create a test framework that captures perf thresholds and reports the results to std out.  The current target looks to be for CI but maybe it can be used by developers testing code locally too?
2) Collect metrics for Kubernetes API calls - https://github.com/kubevirt/kubevirt/issues/5732
3) Collect metrics for VM pod mem/cpu - https://github.com/kubevirt/kubevirt/issues/5732
4) Define an API for different types of reporting

Maybe we can start by creating some issue for 1) and 4)

Reply:
>1. Create a test framework that captures perf thresholds and reports the results to std out. The current target looks to be for CI but maybe it can be used by developers testing code locally too?

That is right. Note that, for the CI standpoint, the test will fail if the latency is not satisfied.

Reply:
> 2. Collect metrics for Kubernetes API calls - #5732

Actually, the test focus on collecting the latency of KubeVirt API calls
For example, the following command returns the latency of the following APIs:
```
histogram_quantile(0.99, 
    sum(rest_client_request_duration_seconds_bucket{url=~".*kubevirt.*"}) by(url,verb,le)
)
```
```
{url="https://virt-api.kubevirt.svc:443/virtualmachineinstances-mutate;verb="POST"}
{url="https://kubevirt-operator-webhook.kubevirt.svc:443/kubevirt-validate-update;verb="POST"}
{url="https://virt-api.kubevirt.svc:443/virtualmachineinstances-validate-update;verb="POST"}
{url="https://kubevirt-operator-webhook.kubevirt.svc:443/kubevirt-validate-delete;verb="POST"}
{url="https://virt-api.kubevirt.svc:443/virtualmachineinstances-validate-create;verb="POST"}
```
But, of course, without the `url` filter, we can collect for all Kubernetes APIs as well... But, this is not the goal of this PR.

Reply:
> 3. Collect metrics for VM pod mem/cpu - #5732

Yes. 

Also, in my test, I saw a big CPU overhead. 
The aggregated CPU usage in the VMI namespace was the double of the sum of all VMI CPU limits.

Reply:
> 4. Define an API for different types of reporting

Right, I see this as a consequence of implementing the test.

Reply:
>        Collect metrics for Kubernetes API calls - #5732
> Actually, the test focus on collecting the latency of KubeVirt API calls
> For example, the following command returns the latency of the following APIs:
>
I'll add api latency to #5732.